frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -1305,6 +1305,114 @@ const IR::Node* TypeInference::postorder(IR::Concat* expression) {
     return expression;
 }
 
+const IR::Node* TypeInference::postorder(IR::Key* key) {
+    LOG1("TypeInference: visiting " << dbp(key));
+    // compute the type and store it in typeMap
+    auto vec = new IR::Vector<IR::Type>();
+    for (auto ke : *key->keyElements) {
+        auto kt = typeMap->getType(ke->expression);
+        if (kt == nullptr) {
+            LOG2("Bailing out for " << dbp(ke));
+            return key;
+        }
+        vec->push_back(kt);
+    }
+    auto keyTuple = new IR::Type_Tuple(Util::SourceInfo(), vec);
+    LOG2("Setting key type to " << dbp(keyTuple));
+    setType(key, keyTuple);
+    // installing also for the original because we cannot tell which one will survive in the ir
+    // \TODO: figure out the rules of survivorship
```

Comment:
Actually O(log n) space (bounded by the depth of the tree, not the size of the tree). I'm not sure how to design a visitor pattern that does no automatic clones, since you need to clone any node that has any children changed (recursively through grandchildren) even if you don't otherwise want to change the node. When I tried it, you end up needing to manually define preorder and postorder functions for almost every possible IR type, which seems impractical.